### PR TITLE
[ADD] features --nsfilters --auto_generate and --subdir

### PR DESCRIPTION
**Pull request by dopch**
Hello,

[ADD] features & support for organization + nameserver filter and subdir organization.
--organization_id TEXT  Gandi organization ID (passing an organization ID will limit query on domain own by organization)

--list_domains          Fetch and display fqdn_unicodde
--list_domains_details  Fetch and display all domain details
--list_organizations    Fetch and display all organizations details

--nsfilters TEXT        Filter domaine based on their current nameserver 'abc','livedns' or 'other'. You can add multiple filter Example: gandi2tf --list_domains --nsfilters abc --nsfilters livedns)

--auto_generate         Auto-generate tf resource based on domain from gandi instead of STDIN/USERINPUT (option can be used with --organization_id and multiple --nsfilters)

--subdir Create a sub-directory to store generated domain tf and tf import command

### CODE REVIEW

**Comment by marcaurele** [comment]
Thanks @dopch for the PR and improvements. I like the change about the sub directory and the filtering based on organizations. On the other side, I don't understand the need for the printing functionality as the tool is not intended to be a cli to visualize the data on Gandi API.

**Comment by dopch** [comment]
Hi @marcaurele,

Sadly https://github.com/Gandi/gandi.cli has now been archived, and in my case i use the printing funcionality because i have to manage +500 domains across multiple orga and filtering and offering display was neccesary for me and other team members.

Sincerely
Dopch

**Comment by dopch** [comment]
Hi @marcaurele,

Commit [https://github.com/dopch/gandi-2-terraform/commit/1e1b7bca100748183be5c8986a0a69641da4bd2d](url), fix a 'bug' if your domain start with a number details on [https://github.com/dopch/gandi-2-terraform/issues/1](url)

File: gandi_tf/main.py [hunk 1]
```diff
@@ -87,33 +148,71 @@ def generate_tf(domain, entries):
...
+                fi.write(f"{cmd}\n")
     return commands
 
 
 @click.command(no_args_is_help=True)
 @click.option("--version", help="Display version information", is_flag=True)
+@click.option("--organization_id", help="Gandi organization ID (passing an organization ID will limit query on domain own by organization)", default=None)
+@click.option("--list_domains", help="Fetch and display fqdn_unicode", is_flag=True)
+@click.option("--list_domains_details", help="Fetch and display all domain details", is_flag=True)
+@click.option("--list_organizations", help="Fetch and display all organizations details", is_flag=True)
```
Comment (marcaurele) [inline gandi_tf/main.py:163]: I still don't want to integrate the cli type of task into this tool, it is not the purpose of it.
```suggestion
```

File: gandi_tf/main.py [hunk 2]
```diff
@@ -87,33 +148,71 @@ def generate_tf(domain, entries):
...
     """
     Command to read Gandi.net live DNS records and generate
     corresponding TF gandi_livedns_record resources.
+
+    Warning: nameserver type 'other' and 'abc' can't be managed via terraform (for abc domain you can transfer them to livedns on gandi webinterface)
     """
+    for filter in nsfilters:
+        if filter not in ['abc','livedns','other']:
+            print(f"nsfilters error available filter: 'abc', 'livedns' or 'other' you submitted {nsfilters}")
+            return
```
Comment (marcaurele) [inline gandi_tf/main.py:179]: Please read the documentation on [click](https://click.palletsprojects.com) and more precisely how to provide [choice options](https://click.palletsprojects.com/en/8.1.x/options/#choice-options) for the parameter. All this code would be unneeded then.

File: gandi_tf/main.py [hunk 3]
```diff
@@ -87,33 +148,71 @@ def generate_tf(domain, entries):
...
-
+    if list_domains:
+        print_domains_list(fetch_domains_list(organization_id), list_domains_details, nsfilters)
+        return
+    if list_domains_details:
+        print_domains_list(fetch_domains_list(organization_id), list_domains_details, nsfilters)
+        return
+    if list_organizations:
+        print_organizations_list(fetch_organizations_list())
+        return
```
Comment (marcaurele) [inline gandi_tf/main.py:194]: To remove:
```suggestion
```

File: gandi_tf/main.py [hunk 4]
```diff
@@ -87,33 +148,71 @@ def generate_tf(domain, entries):
...
     return commands
 
 
 @click.command(no_args_is_help=True)
 @click.option("--version", help="Display version information", is_flag=True)
+@click.option("--organization_id", help="Gandi organization ID (passing an organization ID will limit query on domain own by organization)", default=None)
+@click.option("--list_domains", help="Fetch and display fqdn_unicode", is_flag=True)
+@click.option("--list_domains_details", help="Fetch and display all domain details", is_flag=True)
+@click.option("--list_organizations", help="Fetch and display all organizations details", is_flag=True)
+@click.option('--nsfilters', help="Filter domaine based on their current nameserver 'abc','livedns' or 'other'. You can add multiple filter Example: gandi2tf --list_domains --nsfilters abc --nsfilters livedns)", multiple=True, default=[
```
Comment (marcaurele) [inline gandi_tf/main.py:164]: Can you point me to more documentation on Gandi website about the definition of those values added as a filter?

File: gandi_tf/main.py [hunk 5]
```diff
@@ -87,33 +148,71 @@ def generate_tf(domain, entries):
...
+    if list_organizations:
+        print_organizations_list(fetch_organizations_list())
+        return
+    if auto_generate:
+        domains = []
+        fetch_domains = fetch_domains_list(organization_id)
+        for domain in fetch_domains:
+            if domain['nameserver']['current'] in nsfilters:
+                if "other" in domain['nameserver']['current']:
+                    print("Your are trying to auto_generate a tf from a domain not managed by Gandi check tour nsfilters")
```
Comment (marcaurele) [inline gandi_tf/main.py:201]: There's a typo `...by Gandi check tour nsfilters`. Use also [click.echo](https://click.palletsprojects.com/en/8.1.x/utils/#printing-to-stdout) or [click.secho](https://click.palletsprojects.com/en/8.1.x/utils/#ansi-colors) to output messages. This is a general comment for all the print statements that were added.

File: gandi_tf/main.py [hunk 6]
```diff
@@ -87,33 +148,71 @@ def generate_tf(domain, entries):
...
 
 @click.command(no_args_is_help=True)
 @click.option("--version", help="Display version information", is_flag=True)
+@click.option("--organization_id", help="Gandi organization ID (passing an organization ID will limit query on domain own by organization)", default=None)
+@click.option("--list_domains", help="Fetch and display fqdn_unicode", is_flag=True)
+@click.option("--list_domains_details", help="Fetch and display all domain details", is_flag=True)
+@click.option("--list_organizations", help="Fetch and display all organizations details", is_flag=True)
+@click.option('--nsfilters', help="Filter domaine based on their current nameserver 'abc','livedns' or 'other'. You can add multiple filter Example: gandi2tf --list_domains --nsfilters abc --nsfilters livedns)", multiple=True, default=[
+'abc','livedns','other'])
+@click.option("--auto_generate", help="Auto-generate tf resource based on domain from gandi instead of STDIN/USERINPUT (option can be used with --organization_id, --subdir and multiple --nsfilters)", is_flag=True)
```
Comment (marcaurele) [inline gandi_tf/main.py:166]:
```suggestion
@click.option("--auto-enerate", help="Auto-generate tf resources based on domain from gandi instead of STDIN/USERINPUT (option can be used with --organization-id, --subdir and multiple --nsfilters)", is_flag=True)
```

File: gandi_tf/main.py [hunk 7]
```diff
@@ -87,33 +148,71 @@ def generate_tf(domain, entries):
...
+    if subdir is True:
+        with click.open_file(filename_tfimport, "w") as fi:
+            for cmd in commands:
+                fi.write(f"{cmd}\n")
     return commands
 
 
 @click.command(no_args_is_help=True)
 @click.option("--version", help="Display version information", is_flag=True)
+@click.option("--organization_id", help="Gandi organization ID (passing an organization ID will limit query on domain own by organization)", default=None)
```
Comment (marcaurele) [inline gandi_tf/main.py:160]: Could you turn this to a multi option so people could append multiple `--organization-id` parameters to the command.

```suggestion
@click.option("--organization-id", help="To list domains only owned by this organization id", default=None)
```

File: gandi_tf/main.py [hunk 8]
```diff
@@ -25,6 +23,55 @@ def fetch_records(domain):
     return r.text
 
 
+def fetch_domains_list(organization_id):
+    if organization_id is not None:
+        fh_url = f"https://api.gandi.net/v5/domain/domains?per_page=1&sort_by=fqdn&sharing_id={organization_id}"
```
Comment (marcaurele) [inline gandi_tf/main.py:28]: Please use the functionality of the requests package to send parameters to an URL: https://requests.readthedocs.io/en/latest/user/quickstart/#passing-parameters-in-urls

File: gandi_tf/main.py (hunk 4 above)
Comment (dopch) [reply gandi_tf/main.py:164]: Hi @marcaurele,

Sorry for the delayed responsed i did not had any time assigned to the project

[https://api.gandi.net/docs/domains/](https://api.gandi.net/docs/domains/) GET list domains you will find those filter sadly is a string and we can't pass a list

nameserver ⁠string
One of: "abc", "livedns", "other"
Used to filter the type of nameservers

File: gandi_tf/main.py (hunk 7 above)
Comment (dopch) [reply gandi_tf/main.py:160]: Hi @marcaurele,
organization_id can't be renamed organization-id because "-id" is interpreted has id python builtin.

[https://api.gandi.net/docs/reference/#Sharing-ID](https://api.gandi.net/docs/reference/#Sharing-ID) can be pass as param only has a string an not a list, building a multiple organization_id support will need more work (next release maybe)

File: gandi_tf/main.py (hunk 1 above)
Comment (dopch) [reply gandi_tf/main.py]: Hi @marcaurele,

Done

File: gandi_tf/main.py (hunk 2 above)
Comment (dopch) [reply gandi_tf/main.py]: Hi @marcaurele,

Done

File: gandi_tf/main.py (hunk 3 above)
Comment (dopch) [reply gandi_tf/main.py]: Hi @marcaurele,

Done

File: gandi_tf/main.py (hunk 6 above)
Comment (dopch) [reply gandi_tf/main.py]: Hi @marcaurele,

Done renamed auto-generate

File: gandi_tf/main.py (hunk 5 above)
Comment (dopch) [reply gandi_tf/main.py]: Hi @marcaurele,

Done

**Comment by marcaurele** [comment]
I had a couple of fixes which I added after merging the changes. Thank you @dopch for the changes and extra features. Appreciated!

File: gandi_tf/main.py [hunk 9]
```diff
@@ -87,33 +123,54 @@ def generate_tf(domain, entries):
...
+@click.option("--auto-generate", help="Auto-generate tf resources based on domain from gandi instead of STDIN/USERINPUT (option can be used with --organization_id, --subdir and multiple --nsfilters)", is_flag=True)
+@click.option("--subdir", help="Create a sub-directory to store generated domain tf and tf import command", is_flag=True)
 @click.argument("domains", nargs=-1)
-def generate(domains, version):
+def generate(domains, version, organization_id, nsfilters, auto_generate, subdir):
     """
     Command to read Gandi.net live DNS records and generate
     corresponding TF gandi_livedns_record resources.
+
+    Warning: nameserver type 'other' and 'abc' can't be managed via terraform (for abc domain you can transfer them to livedns on gandi webinterface)
```
Comment (marcaurele) [inline gandi_tf/main.py:145]: Due to my mistake I have to ask question afterwards... Why did you add the option `nsfilters` to contain the non-compatible values?

Comment (dopch) [reply gandi_tf/main.py:145]: Hi @marcaurele 

Well nsfilters was design to work with the auto-generate feature, for example i'm using it to retrieve and generate tf and import state file of all domain available in my organisation that are currently manage via gandi livedns filtering out automatically  domain we own currently delegated to other provider   

Example of command fetching all domain managed by organization XYZ and placing them in subdir:
gandi2tf  --organization_id XYZ --autogenerate --nsfilters livedns --subdir 

After review of the element you chose to merge into your project you can remove reference to nsfilters, organization_id, and fetch_domains_list function and call to function 

If i'm not mistaken you will endup with feature --subdir and the patch for domain starting with number

If you need me to do a cleanup and propose a new version via MR let me know

Sincerely
Dopch

Comment (marcaurele) [reply gandi_tf/main.py:145]: i only have entries with livedns. My understanding of your explanation is that it should only retrieve the _livedns_ entries, the other one should be left out of the TF code. Therefore it should force the `nameserver` paramter to `livedns`.

Comment (marcaurele) [reply gandi_tf/main.py:145]: See https://github.com/marcaurele/gandi-2-terraform/pull/11/commits/d1ff185d2dc69ce0d8134147f6eeb0410a0ffa6d

Comment (dopch) [reply gandi_tf/main.py:145]: fetch with the filter livedns is indeed correct for your project scope
